scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/extension/Conversion.scala

```diff
@@ -0,0 +1,257 @@
+import pl.touk.nussknacker.engine.api.util.ReflectUtils
+import pl.touk.nussknacker.engine.extension.Conversion.toNumber
+import pl.touk.nussknacker.engine.spel.internal.RuntimeConversionHandler
+import pl.touk.nussknacker.engine.util.classes.Extensions.ClassExtensions
+
+import scala.util.{Success, Try}
+import java.util.{
+  ArrayList => JArrayList,
+  Collection => JCollection,
+  HashMap => JHashMap,
+  List => JList,
+  Map => JMap,
+  Set => JSet
+}
+import java.lang.{Boolean => JBoolean, Double => JDouble, Long => JLong, Number => JNumber}
+import java.math.{BigDecimal => JBigDecimal, BigInteger => JBigInteger}
+
+sealed trait Conversion {
+  type ResultType
+  val resultTypeClass: Class[ResultType]
+
+  def canConvert(value: Any): JBoolean
+  def convert(value: Any): ResultType
+  def convertOrNull(value: Any): ResultType
+
+  def typingResult: TypingResult        = Typed.typedClass(resultTypeClass)
+  def simpleSupportedResultTypeName     = ReflectUtils.simpleNameWithoutSuffix(resultTypeClass)
+  def supportedResultTypes: Set[String] = Set(resultTypeClass.getName, simpleSupportedResultTypeName)
+}
+
+object Conversion {
+
+  private[extension] def toNumber(stringOrNumber: Any): JNumber = stringOrNumber match {
+    case s: CharSequence =>
+      val ss = s.toString
+      // we pick the narrowest type as possible to reduce the amount of memory and computations overheads
+      val tries: List[Try[JNumber]] = List(
+        Try(java.lang.Integer.parseInt(ss)),
+        Try(java.lang.Long.parseLong(ss)),
+        Try(java.lang.Double.parseDouble(ss)),
+        Try(new JBigDecimal(ss)),
+        Try(new JBigInteger(ss))
+      )
+
+      tries
+        .collectFirst { case Success(value) =>
+          value
+        }
+        .getOrElse(new JBigDecimal(ss))
+
+    case n: JNumber => n
+  }
+
+}
+
+object LongConversion extends Conversion {
+  override type ResultType = JLong
+  override val resultTypeClass: Class[JLong] = classOf[JLong]
+
+  override def canConvert(value: Any): JBoolean = toLongEither(value).isRight
+
+  override def convert(value: Any): JLong = toLongEither(value) match {
+    case Right(result) => result
+    case Left(ex)      => throw ex
+  }
+
+  override def convertOrNull(value: Any): JLong = toLongEither(value) match {
+    case Right(result) => result
+    case Left(_)       => null
+  }
+
+  private def toLongEither(value: Any): Either[Throwable, JLong] = {
+    value match {
+      case v: String => Try(JLong.valueOf(toNumber(v).longValue())).toEither
+      case v: Number => Right(v.longValue())
+      case _         => Left(new IllegalArgumentException(s"Cannot convert: $value to Long"))
+    }
+  }
+
+}
+
+object DoubleConversion extends Conversion {
+  override type ResultType = JDouble
+  override val resultTypeClass: Class[JDouble] = classOf[JDouble]
+
+  override def canConvert(value: Any): JBoolean = toDoubleEither(value).isRight
```

Comment:
It's because of DRY. For example in `NumericConversionExt` we have methods: `isLong`, `toLong` and `toLongOrNull` and in `ConversionExt` we want to use the same implementation for `to('Boolean')`. I'm wondering if I understand this comment correctly or it's related to sth other?

Reply:
I see that in every Conversion class you do the same in all three methods - only toXYZEither is different. It looks like we can have it done once and have this one method implemented differently. Am I right?